### ISSUE TITLE
NullReferenceException in PopulateRequestAttributes

### ISSUE THREAD

**Issue by mbp** [aws/aws-xray-sdk-dotnet#64]
In ASP.NET Core (2.2, but earlier versions also apply), you can get a NullReferenceException for requests that do not contain the host header:

```
Object reference not set to an instance of an object.

 at Microsoft.AspNetCore.Http.Extensions.UriHelper.GetDisplayUrl(HttpRequest request)
   at Amazon.XRay.Recorder.Handlers.AspNetCore.Internal.AWSXRayMiddleware.PopulateRequestAttributes(HttpRequest request, Dictionary`2 requestAttributes)
   at Amazon.XRay.Recorder.Handlers.AspNetCore.Internal.AWSXRayMiddleware.ProcessHTTPRequest(HttpContext context)
   at Amazon.XRay.Recorder.Handlers.AspNetCore.Internal.AWSXRayMiddleware.Invoke(HttpContext context)
```

Related issue: https://github.com/aspnet/AspNetCore/issues/2718

This can be solved by implementing your own GetDisplayUrl variant, which does the null check. You can basically just copy the one which will be used in .NET Core 3.0: https://github.com/aspnet/AspNetCore/commit/187e89f6f0ee67728a50f987d63234c1900130f

**Comment by yogiraj07** [comment]
Hi @mbp ,
From the error stack, it looks like the issue is [here](https://github.com/aws/aws-xray-sdk-dotnet/blob/master/sdk/src/Handlers/AspNetCore/Internal/AWSXRayMiddleware.cs#L278). Thank you for the detail analysis.

It seems, for .NET Core, we need to implement our own logic (null safe) to get URL from the incoming request.

I will look into this soon. A PR is also welcomed and would be happy to review it. 

Best,
Yogi

**Comment by yogiraj07** [comment]
Hi @mbp ,
Can you help me in reproducing this issue?
I tried spinning up ASP.NET Core 2.2 app locally, instrumented with X-Ray and curled the endpoint with following command:
`curl -v --http1.0 -H "Host:" <address>`

However, the server doesn't accept request with null host value and returns following response:

```
DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request - Invalid Hostname</h2>
<hr><p>HTTP Error 400. The request hostname is invalid.</p>
</BODY></HTML>

```

As a result, the request never reached the `GetDisplayUrl` code path of the X-Ray middleware.

Few details about the reproduction, the stack you used will be helpful to fix the issue.

Thanks,
Yogi

**Comment by mbp** [comment]
I believe you are getting that error page from IIS Express because it will only listen for requests for hostname = localhost

When you have created ASP.NET Core Web Application in Visual Studio, there is a dropdown in where you can change it from "IIS Express" to "WebApplication", so it runs directly with Kestrel instead of via IIS Express

![image](https://user-images.githubusercontent.com/77322/55880688-87ef9b00-5ba1-11e9-978c-8d8572391928.png)

I was able to reproduce this in a http scenario (not https)

**Comment by yogiraj07** [comment]
Hi @mbp ,
 I could reproduce the issue locally. I wanted to check with Net Core team whether they are planning to fix the issue in 2.0 and will have same behavior as NET Core 3.0. I have commented on the [issue](https://github.com/aspnet/AspNetCore/issues/2718). 

I would try to avoid copy pasting code at the moment and will wait for the reply on the other thread. I have also pointed out 2 options to solve this issue on the same thread.

Thanks for the patience.

Yogi.
